runtime/src/iree/hal/buffer.c

```diff
@@ -21,70 +21,101 @@
 // String utils
 //===----------------------------------------------------------------------===//
 
+static const iree_bitfield_string_mapping_t iree_hal_memory_type_mappings[] = {
+    // Combined:
+    {IREE_HAL_MEMORY_TYPE_HOST_LOCAL, IREE_SVL("HOST_LOCAL")},
+    {IREE_HAL_MEMORY_TYPE_DEVICE_LOCAL, IREE_SVL("DEVICE_LOCAL")},
+    // Separate:
+    {IREE_HAL_MEMORY_TYPE_OPTIMAL, IREE_SVL("OPTIMAL")},
+    {IREE_HAL_MEMORY_TYPE_HOST_VISIBLE, IREE_SVL("HOST_VISIBLE")},
+    {IREE_HAL_MEMORY_TYPE_HOST_COHERENT, IREE_SVL("HOST_COHERENT")},
+    {IREE_HAL_MEMORY_TYPE_HOST_CACHED, IREE_SVL("HOST_CACHED")},
+    {IREE_HAL_MEMORY_TYPE_DEVICE_VISIBLE, IREE_SVL("DEVICE_VISIBLE")},
+};
+
+IREE_API_EXPORT iree_status_t iree_hal_memory_type_parse(
+    iree_string_view_t value, iree_hal_memory_type_t* out_value) {
+  return iree_bitfield_parse(value,
+                             IREE_ARRAYSIZE(iree_hal_memory_type_mappings),
+                             iree_hal_memory_type_mappings, out_value);
+}
```

Comment:
Do these strings ("DEVICE_VISIBLE", etc.) end up in the final program binary if the `_parse` methods are never called? The parsing certainly looks useful, but we could save quite a few bytes if they can be stripped in binary size-optimized configs.

Reply:
nope! they're static to the file so if DCE is able to remove the functions using them in here (the parse/format) they'll be dropped as well - the nice thing with this util is that if both are used we still only have one copy of the strings instead of two for the bidirectional mapping